main.go

```diff
@@ -0,0 +1,21 @@
+package main
+
+import (
+	"strconv"
+)
+
+func knapsack(w int, val []int, wt []int, cache map[string]int) (max int) {
+	if v, ok := cache[strconv.Itoa(w)+" "+strconv.Itoa(len(val))]; ok {
```

Comment:
Converting to string and comparing strings are costly operations. It's more efficient to use struct `struct {w int, v int}` as a map key.